[CL-2909] Refactoring Dockerfile

adessy opened the pull request:
## Changelog
- Pin versions of apt-get package dependencies in Dockerfile.
- Delete Bundler cache to reduce the size of the Docker image.
- Add a build argument to the Dockerfile to allow to tuning the number of jobs used by `bundle install`.

adessy committed:
[CL-2909] Use binstubbed version of rails/rake when running rake tasks

adessy committed:
[CL-2909] Exclude vendor directory from Rubocop scope

adessy committed:
[CL-2909] Add Dockerfile ARG to set the --jobs option of bundle install

adessy committed:
[CL-2909] Delete bundler cache

Disable bundler cache to decrease the size of the image.

The Bundler cache is a directory that stores a copy of each gem that is installed by Bundler. When Bundler installs a gem, it downloads the gem's files and stores them in the cache directory. If the gem is already in the cache, Bundler will use the cached copy instead of downloading it again.

adessy committed:
[CL-2909] Pin versions in apt get install.

adessy committed:
Merge branch 'master' into CL-2909/optimize-production-dockerfile-2

alexander-cit commented on back/Dockerfile:
```diff
@@ -20,35 +20,45 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
- && apt-get -y --no-install-recommends install nodejs            \
+RUN apt-get update                                                             \
+ && apt-get install -qq -y --no-install-recommends                             \
+      build-essential=12.9                                                     \
```
Specifying exact versions in apt doesn't seem to be common, does it? What will happen if some of these packages get security updates? I suppose we'll not be notified about it, so not sure we should do it. But up to you.

adessy replied:
It's actually recommended (but often overlooked). You're right that it might introduce some extra overhead though. Will revert it for now.

alexander-cit commented on back/Dockerfile:
```diff
@@ -20,35 +20,45 @@ SHELL ["/bin/bash", "-o", "pipefail", "-c"]
+ARG BUNDLER_JOBS
+RUN bundle config set --local clean 'true'               \
+ && bundle config set --local deployment 'true'          \
+ && bundle install ${BUNDLER_JOBS:+--jobs=$BUNDLER_JOBS} \
```
It looks like there's already `BUNDLE_JOBS` env var https://bundler.io/v2.4/man/bundle-config.1.html#:~:text=The%20number%20of%20gems%20Bundler%20can

I also thought about giving it a default of 4 (which would be useful for medium resources), but the current default makes sense too.

adessy replied:
Nice, I didn't know about that one. Will simplify the code.

I'd rather not set a default value to avoid unnecessary coupling between the Dockerfile definition and the executor.

alexander-cit replied:
Makes sense.

alexander-cit reviewed (approved):
No blockers.

adessy committed:
[CL-2909] Unpin versions of apt-get packages

It would be better to have them pinned, but we would have to monitor security updates and keep them up-to-date manually.

adessy committed:
[CL-2909] Rename BUNDLER_JOBS Dockerfile argument to BUNDLE_JOBS

BUNDLE_JOBS is automatically picked up by bundler. So it simplifies the code a bit.
